build(deps): Bump from vue 2 to vue 3

shariquerik opened the pull request:
- Upgrade `vue 2.6.14` to `vue 3.2.39`
- Upgrade `vue-router 2.0.0` to `vue-router 4.1.5`
- Upgrade `vuex 3` to `vuex 4.0.2`
- Upgrade `vuedraggable` from `2.24.3` to `4.1.0`
- Using `esbuild-plugin-vue3` for building vue 3.x SFC files
- Written `frappe_vue_style` plugin to prepend the `frappe.require("<file-name>.bundle.css")` to bundle.js file (Thanks @netchampfaris for help). In vue 2 CSS was injected in the js file in vue 3 two files are getting build js and CSS. Also had to include the mapping of CSS file in `assets.json`
- Made some changes to support `frappe.require()` on the website.
- Using Composition API
- Removed `module_home.js`. It is not used anywhere and it does not do anything.

**Vue projects to refactor to support vue 3:**
- [x] Recorder
- [x] File Uploader
- [x] Print Format Builder Beta
- [x] Build Event
- [x] Kanban Board using Vuex

**Some points to know while working with vue 3 implementation:**
1. Was not able to use `__("Translatable String")` in template HTML. It is not working because `Vue.prototype` is deprecated now which was used to make `window.__` accessible in vue code. Implemented the recommended solution https://v3-migration.vuejs.org/breaking-changes/global-api.html#vue-prototype-replaced-by-config-globalproperties
2. `ref()` is working but `reactive()` is not working. Not sure why. (reactive again uses ref so using ref instead)(Not working only in recorder, I tested separately it works)
3. I saved the vue instance in variable `frappe.recorder.view` now in vue3 you will not be able to get $route or $router from this instance since now we get route & router from vue-router. I save the route & router in `frappe.recorder.route(r)` which solved the problem. Maybe this can be solved by using state management.
4. We can no longer send objects as params to the router while switching routes. So I used `router.BeforeEach` hook to set a meta value which I use later as params. https://router.vuejs.org/guide/advanced/meta.html#route-meta-fields
5. In vue 3, not all the methods and variables are accessible to parents by default everything is restricted if you want to expose them use `defineExpose`.
6. For vuedraggable check the migration guide from https://github.com/SortableJS/vue.draggable.next
7. Mixin was removed you can replace it with Composables.  https://vuejs.org/guide/reusability/composables.html

**Need Suggestion:**
Since the way in which vue is initialized has changed we used to initialize Vue at the global level and set vue globals only once like this in vue 2 `Vue.prototype.<globalvariable> = window.<variable>` in vue 3 we cannot use same Vue instance across multiple apps, for each app we have to initialize new vue instance. At the same time, we will have to set vue globals. 
Because of this bundle size is also increased.

shariquerik committed:
build(deps): Bump from vue 2 to vue 3

Alchez commented:
@shariquerik, this change is very welcome!

I have a couple of notes/questions:
- The `@vue/compiler-sfc` dependency is [now available within the main `vue` package](https://www.npmjs.com/package/@vue/compiler-sfc), so you shouldn't need to explicitly use the lib.
- Any plans to use [Pinia](https://pinia.vuejs.org/) instead of Vuex, since that's what [Vue recommends](https://vuex.vuejs.org/) using now?
- Anything to know if we start contributing on the refactor projects you mentioned?

shariquerik committed:
chore: linter fixes

shariquerik committed:
chore: removed compiler-sfc from package.json since it comes with vue

shariquerik commented:
> * The `@vue/compiler-sfc` dependency is [now available within the main `vue` package](https://www.npmjs.com/package/@vue/compiler-sfc), so you shouldn't need to explicitly use the lib.

Did it thanks👍🏻

> * Any plans to use [Pinia](https://pinia.vuejs.org/) instead of Vuex, since that's what [Vue recommends](https://vuex.vuejs.org/) using now?

Currently, in this PR I am not planning to maybe in separate PR. Or feel free to contribute.

> * Anything to know if we start contributing on the refactor projects you mentioned?

Not started refactoring yet, will keep updating the description.

shariquerik committed:
refactor: Recorder code to vue3

shariquerik committed:
chore: linter fixes

shariquerik committed:
refactor: Build Event code to vue3

shariquerik committed:
refactor: updated vuex syntax in Kanabn Board

shariquerik committed:
chore: removed module_home.js

unused code

shariquerik committed:
refactor: updated vue syntax

shariquerik committed:
fix: set vue global properties

since Vue.prototype is removed using app.config.globalProperties to set window.__ & window.frappe

shariquerik committed:
fix: removed Vue.prototype implementation

nextchamp-saqib commented:
> * Any plans to use [Pinia](https://pinia.vuejs.org/) instead of Vuex, since that's what [Vue recommends](https://vuex.vuejs.org/) using now?

Vuex is used only for Kanban Board and it doesn't seem to be complex enough to use a state management library
Vue 3's Composition API provides all the necessary blocks for managing state just like Vuex
In my opinion, we should use Composition API (but yes, later)

Alchez commented:
> Vuex is used only for Kanban Board and it doesn't seem to be complex enough to use a state management library
Vue 3's Composition API provides all the necessary blocks for managing state just like Vuex
In my opinion, we should use Composition API (but yes, later)

Yeah, that makes sense. A more general rework of the Frappe views might require it, but SFCs should be fine for now.

shariquerik committed:
fix: handle loading of bundle css & js in website

shariquerik committed:
Merge branch 'develop' into migrate-to-vue3

shariquerik committed:
fix: removed unused script

shariquerik committed:
refactor: File Uploader code to vue 3

shariquerik committed:
refactor: Print Format Builder Beta code to vue 3

shariquerik committed:
fix: get assets_json from boot

shariquerik committed:
chore: minor fix

shariquerik committed:
Merge branch 'develop' into migrate-to-vue3

shariquerik commented:
Unrelated test failing

shariquerik committed:
Merge branch 'develop' into migrate-to-vue3

shariquerik committed:
Merge branch 'develop' into migrate-to-vue3

sagarvora commented:
looks like I'll finally need to branch India Compliance into `version-14` and <s>develop</s>`next` 😅 

edit: <s>or maybe I can use vue2 by specifying it?</s> this doesn't work because of the way dependencies are compiled when building JS assets in `esbuild.js`
